File: Programmers/연습문제/Level 2/문자열 압축/Solution.js
function solution(s) {
    let answer = s.length;
    for (let length=1; length<s.length/2+1; length++) {
        let slice = getSlice(s, length);
        let zippedLength = getZippedLength(slice);
        if (answer > zippedLength) {
            answer = zippedLength;
        }
    }
    return answer;
}

function getSlice(s, length) {
    let pattern = new RegExp('.{1,' + length + '}', 'g');
    return s.match(pattern);
}

function getZippedLength(slice) {
    let zipped = ''
    let lastWord = '';
    let count = NaN;
    for (const word of slice) {
        if (word != lastWord) {
            zipped += count >= 2 ? count : '';
            zipped += lastWord;
            count = 1;
        } else {
            count++;
        }
        lastWord = word;
    }
    zipped += count >= 2 ? count : '';
    zipped += lastWord;
    return zipped.length;
}


// TC
const s = "ababcdcdababcdcd";
console.log(solution(s));

File: Programmers/연습문제/Level 2/문자열 압축/Solution.swift
import Foundation

func solution(_ s:String) -> Int {
    if s.count == 1 { return 1 }
    
    var answer = s.count
    for length in (1...s.count/2) {
        let zipped_s = zipped(s, length)
        answer = min(answer, zipped_s.count)
    }
    return answer
}

func zipped(_ s: String, _ length: Int) -> String {
    var zippedResult = ""
    var zipWord = ""
    var zipCount = 1
    for i in stride(from: 0, to: s.count, by: length) {
        let startIndex = String.Index(utf16Offset: i, in: s)
        let endIndex = String.Index(utf16Offset: min(i+length, s.count), in: s)
        let word = String(s[startIndex..<endIndex])
        if word == zipWord {
            zipCount += 1
        } else {
            zippedResult += (zipCount > 1 ? String(zipCount) : "") + zipWord
            zipWord = word
            zipCount = 1
        }
    }
    zippedResult += (zipCount > 1 ? String(zipCount) : "") + zipWord
    return zippedResult
}


// TC
let s = "aabbaccc"
print(solution(s)) // 7
